ktor-network/ktor-network-tls/ktor-network-tls-certificates/jvm/src/io/ktor/network/tls/certificates/builders.kt

```diff
@@ -47,38 +57,88 @@ public class CertificateBuilder internal constructor() {
      */
     public var keySizeInBits: Int = 1024
 
+    /**
+     * The usage for the generated key.
+     *
+     * This determines the extensions that should be written in the certificate, such as [OID.ExtKeyUsage] (server or
+     * client authentication), or [OID.BasicConstraints] to use the key as CA.
+     */
+    public var keyType: KeyType = KeyType.Server
+
+    /**
+     * Domains for which this certificate is valid (only relevant for [KeyType.Server], ignored for other key types).
+     */
+    public var domains: List<String> = listOf("localhost")
+
+    /**
+     * IP addresses for which this certificate is valid (only relevant for [KeyType.Server], ignored for other key types).
+     */
+    public var ipAddresses: List<InetAddress> = listOf(Inet4Address.getByName("127.0.0.1"))
+
+    private var issuer: CertificateIssuer? = null
+
+    private data class CertificateIssuer(
+        val name: X500Principal,
+        val keyPair: KeyPair,
+        val keyCertificate: Certificate,
+    )
+
+    /**
+     * Defines an issuer for this certificate, so it is not self-signed.
+     *
+     * The certificate will be signed with the given [issuerKeyPair], certified by the given [issuerKeyCertificate].
+     * The issuer's name is taken from the provided certificate.
+     *
+     * If this method is not called, this certificate will be self-signed by the subject (with the same generated key
+     * as the one being certified).
+     */
+    public fun signWith(
+        issuerKeyPair: KeyPair,
+        issuerKeyCertificate: X509Certificate,
+    ) {
+        issuer = CertificateIssuer(
+            // the subject of the given certificate is the issuer of the certificate we're creating
+            name = issuerKeyCertificate.subjectX500Principal,
+            keyPair = issuerKeyPair,
+            keyCertificate = issuerKeyCertificate,
+        )
+    }
+
+    /**
+     * Defines an issuer for this certificate, so it is not self-signed.
+     *
+     * The certificate will be signed with the given [issuerKeyPair] in the name of [issuerName], certified by the
+     * given [issuerKeyCertificate].
+     *
+     * If this method is not called, this certificate will be self-signed by the subject (with the same generated key
+     * as the one being certified).
+     */
+    public fun signWith(
+        issuerKeyPair: KeyPair,
+        issuerKeyCertificate: Certificate,
+        issuerName: X500Principal,
+    ) {
```

Comment:
This could also just be named `issuer`, with params like `issuer(name: X500Principal, keyPair: KeyPair, keyCertificate: Certificate)`. I just felt like the name `issuer` feels weird for the other overload where the name of the issuer is taken from the certificate. Maybe we don't need that other overload at all 🤷‍♂️